### DISCUSSION TITLE
Experience when binding ISPC functions to Python

### DISCUSSION THREAD

**Discussion by maedoc** [ispc/ispc#2898] [Ideas]
ISPC generates nice header files.  I'm using a Python package `ctypesgen` to generate uncompiled bindings to the functions that my ISPC files `export`, with the caveat that the NumPy arrays passed to ISPC functions need to handled appropriately.  Since the overhead with a compiled binding like nanobind (pybind11's "successor") is lower (4x) I also tried with that, but one needs to redefine all the fields etc; however, an advantage of nanobind is that it anticipates interop with machine learning libraries like PyTorch or JAX.  In other words, one could dispatch a PyTorch op via nanobind to an ISPC implemented function if on CPU or to CUDA/HIP/OpenCL if GPU.  That seems like a big win over plain C/C++. 

I'm curious about thoughts & experience of others!

*I searched for Python in the issues & discussions before posting, sorry if I missed anything*

**Comment by aneshlya** [comment]
Thank you for sharing this! It sounds like nanobind is indeed the way to go, especially considering its lower overhead and interoperability with machine learning libraries.

While we don't have much experience with Python bindings ourselves, we're eager to support you in any way we can. We'd love to try ISPC in machine learning libs but don't have much resources to do so. 
Could you let us know what we can provide to make your process easier? It it possible for you to share some of your experiments so we can try it on our side as well?

**Reply by maedoc** [reply]
thanks for the warm reply.

> Could you let us know what we can provide to make your process easier? 

The header output of ISPC looks already quite nice, and `ctypesgen` is able to parse and convert to JSON, which could be used to automate the nanobind wrapper.

But to brainstorm a bit, one potential improvement would be the ability to interact with the [dlpack format](https://github.com/dmlc/dlpack/blob/main/include/dlpack/dlpack.h#L192), which is the backing format used by nanobind and the other ML frameworks to zero-copy exchange data. With minor modifications to that header, I was able to include it in an ISPC file and compile without complaints, here's the changes <details>

```diff
@@ -32,8 +32,14 @@
 #define DLPACK_DLL
 #endif
 
-#include <stdint.h>
-#include <stddef.h>
+// #include <stdint.h>
+// #include <stddef.h>
+typedef int32 int32_t;
+typedef uint8 uint8_t;
+typedef uint16 uint16_t;
+typedef uint32 uint32_t;
+typedef uint64 uint64_t;
+typedef int64 int64_t;
 
 #ifdef __cplusplus
 extern "C" {
@@ -251,7 +257,7 @@ typedef struct {
  *
  * \sa DLManagedTensorVersioned
  */
-typedef struct DLManagedTensor {
+typedef struct {
   /*! \brief DLTensor which is being memory managed */
   DLTensor dl_tensor;
   /*! \brief the context of the original host framework of DLManagedTensor in
@@ -264,7 +270,7 @@ typedef struct DLManagedTensor {
    * NULL if there is no way for the caller to provide a reasonable destructor.
    * The destructor deletes the argument self as well.
    */
-  void (*deleter)(struct DLManagedTensor * self);
+  void (*deleter)(void * self);
 } DLManagedTensor;
```
The change to struct was because otherwise ISPC says
```
include/dlpack/dlpack.h:260:9: Error: Multiple types provided for 
        declaration. 
   */
  void (*deleter)(void * self);
} DLManagedTensor;
```

</details> 
Being able to use the DLPack structure directly means that the ISPC kernels can "just" take multiple tensors as arguments and get to work.  That would also make the binding aspects easier.  If you have any tips on this, it would be nice to hear.  Since Intel GPUs are part of the ISPC story, and DLPack can reference oneAPI tensors, I guess it could also be interesting to test that out.

If, again brainstorming, I were to go further, I'd want to have easier n-dimensional access to the arrays.  However, I usually notice that specific offsets can be computed once, when I write the linear indexing by hand.  This is fuzzy territory for me, but perhaps ISPC could better optimize the code if it has the DLPack information about strides, shape, etc.  There's probably some lessons learned in the XLA (also LLVM based) codebase? 

> It it possible for you to share some of your experiments so we can try it on our side as well?

I've started working on integrating some experiments into a brain simulator that I work on; the only kernel there right now, computes time delayed interactions.  This particular kernel is onerous to implement with tensor operations, and only performs well with an explicit SIMD programming model like ISPC or CUDA.

https://github.com/maedoc/tvb-root/tree/0670880b11ccf64797f73745f7a403da20e7f755/tvb_kernels

There is the ISPC implementation, with different variants that use different data access patterns, and in the nanobind binding `tvb_kernels.cpp` also a plain CPP implementation.   The "nop" variants help measure binding overhead from Python.

I don't have clean implementations yet for all of these, but more kernels which seem ideal for ISPC on CPU (i.e. low latency, high throughput, guaranteed SIMD) include many recursive algorithms such as

- S/D/ODE integrators (including neural ODEs) with small systems (i.e. working data <= L3 cache size)
- Kalman filters and variants
- Convolutions and state space models
- Normalizing flows ([cf](https://blog.evjang.com/2018/01/nf2.html))

**Reply by nurmukhametov** [reply]
> ```diff
> 
> -typedef struct DLManagedTensor {
> +typedef struct {
>    /*! \brief DLTensor which is being memory managed */
>    DLTensor dl_tensor;
>    /*! \brief the context of the original host framework of DLManagedTensor in
> @@ -264,7 +270,7 @@ typedef struct DLManagedTensor {
>     * NULL if there is no way for the caller to provide a reasonable destructor.
>     * The destructor deletes the argument self as well.
>     */
> -  void (*deleter)(struct DLManagedTensor * self);
> +  void (*deleter)(void * self);
>  } DLManagedTensor;
> ```
> 
> The change to struct was because otherwise ISPC says
> 
> ```
> include/dlpack/dlpack.h:260:9: Error: Multiple types provided for 
>         declaration. 
>    */
>   void (*deleter)(void * self);
> } DLManagedTensor;
> ```
> 

It is same issue as https://github.com/ispc/ispc/issues/2351

**Comment by iperov** [comment]
i don't use bindings or generators.
it's not hard to write a function once in python to use it via ctypes
I use python annotations for ctypes argtypes

example
```python
@dll_import(dll_filepath)
def c_circle_faded(img_out : c_float_p, W : c_int32, H : c_int32, cx : c_float, cy : c_float, fs : c_float, fe : c_float) -> None: ...
```

```python

import ctypes
from ctypes import c_bool, c_float, c_int32, c_void_p
from ctypes.util import find_library
from typing import get_type_hints

dlls_by_name = {}
def dll_import(dll_name):
    """
    decorator for import DLL func.
    always annotate return type even if it is None !
    """
    dll_name = str(dll_name)

    def decorator(func):
        dll_func = None
        def wrapper(*args):
            nonlocal dll_func

            if dll_func is None:
                dll = dlls_by_name.get(dll_name, None)
                if dll is None:
                    try:
                        dll = ctypes.cdll.LoadLibrary(find_library(dll_name))
                    except:
                        pass

                    if dll is None:
                        raise RuntimeError(f'Unable to load {dll_name} library.')

                    dlls_by_name[dll_name] = dll

                dll_func = getattr(dll, func.__name__)
                anno = list(get_type_hints(func).values())
                anno = [ None if x is type(None) else x for x in anno ]

                dll_func.argtypes = anno[:-1]
                dll_func.restype = anno[-1]

            return dll_func(*args)

        return wrapper
    return decorator
```

**Reply by maedoc** [reply]
I like the idea of adding python typing for this. Very cool. 

 I still like ctypesgen ability to generate this based on the ISPC-generated header because it’s automatic (less error prone) two steps from ISPC file. 

Nevertheless there’s some overhead to ctypes not present for other

**Reply by iperov** [reply]
> because it’s automatic (less error prone)

what's your workflow?
usually you program a function for python, test it in python too. In this case, writing annotations corresponding to the C signature is easy and takes very little time compared to writing and testing the function itself.

> overhead to ctypes

ctypes has no overhead. It is direct C dll call. Simple, convenient, pure python, without 3rd party libs dependencies.

**Reply by maedoc** [reply]
> workflow?

I write a NumPy implementation, write a C struct def to hold all the arrays and dimensions, implement the ISPC version with a few variants, run the ctypesgen to create the binding and then use `np.testing.assert_allclose` to check the outputs.

Using structs instead of long-ish argument lists helps reduce overhead of the Python call, but then one has to take care that the argument order is maintained correctly from the implementation to the ctypes usage, which is the gain for me using ctypesgen.  here's an example 

https://github.com/maedoc/tvb-root/blob/0670880b11ccf64797f73745f7a403da20e7f755/tvb_kernels/nodes.ispc#L32

there are a few structs and for this case many more to add, so I appreciate not retyping everything on the Python side.

> no overhead

on my x86_64 machine, a simple Python function `lambda i: i` takes about ~35 ns, while a ctypes call to `void f(int i) {return i;}` takes ~230ns, while the nanobind version takes ~60ns.  it's admittedly a small overhead but not zero.

**Reply by iperov** [reply]
That's the point, you don't need to call the function often. The function is called 1 time, those unfortunate 170ns of overhead are spent, then the function takes control and calculates a large amount of data. If you call the function frequently, then python itself will impose overhead on the release/acquire GIL. 
I wouldn't clutter the project with additional dependencies to build binds.

**Comment by aneshlya** [comment]
Inspired by this discussion, we added examples demonstrating ISPC calls from Python using nanobind and ctypes: https://github.com/ispc/ispc/pull/3299. Nanobind provides a great performance and clean integration so we may consider generating nanobind wrappers for export functions: https://github.com/ispc/ispc/issues/3301

**Comment by nurmukhametov** [comment]
@maedoc  In #3391, we added support for the `--nanobind=<filename>` option, which can be used to generate nanobind wrappers for exported ISPC functions. An example of its usage can be find in the [point_transform](https://github.com/ispc/ispc/tree/main/examples/cpu/point_transform_nanobind) example. You can use the ISPC trunk build (available [here](https://github.com/ispc/ispc/releases/tag/trunk-artifacts)) if you want to test it for your use cases.
